* https://huggingface.co/learn/nlp-course/chapter1/4?fw=pt

* https://www.youtube.com/watch?v=MUqNwgPjJvQ&list=PLo2EIpI_JMQtNtKNFFSMNIZwspj8H7-sQ&index=6
* https://nlp.seas.harvard.edu/2018/04/03/attention.html
* https://jalammar.github.io/visualizing-neural-machine-translation-mechanics-of-seq2seq-models-with-attention/
* https://jalammar.github.io/illustrated-transformer/
* https://jalammar.github.io/illustrated-gpt2/

### pipeline
* **It supports most of the common NLP tasks**
* Text classifier
* Zero-Shot classification
* Text generation
* Text completion(mask filling)
* ner(named entity recognition)
* Token claffication
* Question answering
* Summerization
* Translation

### pipeline
The most basic object in the 🤗 Transformers library is the ***pipeline() function***. It connects a model with its necessary preprocessing and postprocessing steps, allowing us to directly input any text and get an intelligible answer:
[pipeline](./huggingface_pipeline.png)
<img src="./huggingface_pipeline.png" width="800" height="400">
* **some of available pipeline**
    * ```feature-extraction```
    * ```fill-mask```
    * ```ner```(named entity recognition)
    * ```question-answering```
    * ```sentiment-analysis```
    * ```summarization```
    * ```text-generation```
    * ```translation```
    * ```zero-shot-classification```

In [1]:
from transformers import pipeline

C:\Users\YUI2KOR\BOSCH\KNOWLEDGE\NOTEBOOK\env_notebook\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
classifier = pipeline("sentiment-analysis")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
C:\Users\YUI2KOR\BOSCH\KNOWLEDGE\NOTEBOOK\env_notebook\lib\site-packages\huggingface_hub\file_download.py:159: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\YUI2KOR\.cache\huggingface\hub\models--distilbert--distilbert-base-uncased-finetuned-sst-2-english. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either

* **warning**

* No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).

* Using a pipeline without specifying a model name and revision in production is not recommended.

* C:\Users\YUI2KOR\BOSCH\KNOWLEDGE\NOTEBOOK\env_notebook\lib\site-packages\huggingface_hub\file_download.py:159: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\YUI2KOR\.cache\huggingface\hub\models--distilbert--distilbert-base-uncased-finetuned-sst-2-english. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


In [3]:
classifier("I've been waiting for a HuggingFace course my whole life.")


[{'label': 'POSITIVE', 'score': 0.9598049521446228}]

In [6]:
classifierll = pipeline("zero-shot-classification")
classifierll("This is a course about the transformers library",
            candicate_labels=['education', 'politics', 'business'])

No model was supplied, defaulted to facebook/bart-large-mnli and revision c626438 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.
Hardware accelerator e.g. GPU is available in the environment, but no `device` argument is passed to the `Pipeline` object. Model will be on CPU.


TypeError: object of type 'NoneType' has no len()

In [ ]:
generator = pipeline('text-generation')
generator('In this course, we will teach you how to')

In [ ]:
#other way by specifying the model
generator = pipeline('text-generation', model='distilgpt2')
generator('In this course, we will teach you about how to',
         max_length=30,
         num_return_sequences=2)

In [7]:
#fill-mask pipeline will predict missing words in a sentence
unmasker = pipeline('fill-mask')
unmasker('This course will teach you about <mask> models.', 
        top_k=2)

No model was supplied, defaulted to distilbert/distilroberta-base and revision ec58a5b (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.
C:\Users\YUI2KOR\BOSCH\KNOWLEDGE\NOTEBOOK\env_notebook\lib\site-packages\huggingface_hub\file_download.py:159: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\YUI2KOR\.cache\huggingface\hub\models--distilbert--distilroberta-base. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see 

[{'score': 0.19418099522590637,
  'token': 30412,
  'token_str': ' mathematical',
  'sequence': 'This course will teach you about mathematical models.'},
 {'score': 0.05562317371368408,
  'token': 38163,
  'token_str': ' computational',
  'sequence': 'This course will teach you about computational models.'}]

In [ ]:
# NER pipeline identifies entities such as persons, organizations or locations in a sentence.
ner = pipelinr('ner', grouped_entities=True)
ner('My name is Indrajeet kumar and I work at bosch at bangalore')

In [ ]:
#translation pipeline french to english
translator = pipeline('translation', model='Helsinki-NLP/opus-mt-fr-en')
translator('Ce cours est produit par Hugging Face.')

In [1]:
from transformers import pipeline

ner = pipeline("ner", grouped_entities=True)
ner("My name is Sylvain and I work at Hugging Face in Brooklyn.")

No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision f2482bf (https://huggingface.co/dbmdz/bert-large-cased-finetuned-conll03-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/998 [00:00<?, ?B/s]

C:\Users\YUI2KOR\BOSCH\KNOWLEDGE\NOTEBOOK\env_notebook\lib\site-packages\huggingface_hub\file_download.py:159: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\YUI2KOR\.cache\huggingface\hub\models--dbmdz--bert-large-cased-finetuned-conll03-english. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


model.safetensors:   0%|          | 0.00/1.33G [00:00<?, ?B/s]

Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/60.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

C:\Users\YUI2KOR\BOSCH\KNOWLEDGE\NOTEBOOK\env_notebook\lib\site-packages\transformers\tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(
Hardware accelerator e.g. GPU is available in the environment, but no `device` argument is passed to the `Pipeline` object. Model will be on CPU.
C:\Users\YUI2KOR\BOSCH\KNOWLEDGE\NOTEBOOK\env_notebook\lib\site-packages\transformers\pipelines\token_classification.py:168: UserWarning: `grouped_entities` is deprecated and will be removed in version v5.0.0, defaulted to `aggregation_strategy="simple"` instead.
  warnings.warn(


[{'entity_group': 'PER',
  'score': 0.9981694,
  'word': 'Sylvain',
  'start': 11,
  'end': 18},
 {'entity_group': 'ORG',
  'score': 0.9796019,
  'word': 'Hugging Face',
  'start': 33,
  'end': 45},
 {'entity_group': 'LOC',
  'score': 0.9932106,
  'word': 'Brooklyn',
  'start': 49,
  'end': 57}]

### Inside the pipeline
[Source link](https://huggingface.co/learn/nlp-course/chapter2/2?fw=pt).

The pipeline groups together three steps: ```preprocessing | passing the inputs through the model | postprocessing```.
<img src="./huggingface_inside_pipeline.png" width="800" heigh="400">

### 1. preprocessing with tokenizer
* Transformer models can’t process raw text directly, so the first step of our pipeline is to convert the text inputs into numbers. To do this we use a tokenizer, which will be responsible for:

    * Splitting the input into words, subwords, or symbols (like punctuation) that are called tokens
    * Mapping each token to an integer
    * Adding additional inputs that may be useful to the model

**eg.**
Since the default checkpoint of the ```sentiment-analysis pipeline``` is ```distilbert-base-uncased-finetuned-sst-2-english```

In [2]:
from transformers import AutoTokenizer

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

C:\Users\YUI2KOR\BOSCH\KNOWLEDGE\NOTEBOOK\env_notebook\lib\site-packages\huggingface_hub\file_download.py:159: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\YUI2KOR\.cache\huggingface\hub\models--distilbert-base-uncased-finetuned-sst-2-english. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

In [3]:
raw_inputs = [
    "I've been waiting for a HuggingFace course my whole life.",
    "I hate this so much!",
]
inputs = tokenizer(raw_inputs, padding=True, truncation=True, return_tensors="pt")
print(inputs)

{'input_ids': tensor([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662, 12172,
          2607,  2026,  2878,  2166,  1012,   102],
        [  101,  1045,  5223,  2023,  2061,  2172,   999,   102,     0,     0,
             0,     0,     0,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]])}


### 2. Going through the model
* Transformers provides an ```AutoModel``` class which also has a ```from_pretrained()``` method.
* This architecture contains only the base Transformer module: given some inputs, it outputs only hidden states, also known as features. i.e. **contextual understanding of the input by the Transformer model**.
* high-dimensional vector contains:
    * **batch size:** The number of sequence processed at a time(2 in this example)
    * **sequence length:** The length of numerical representation of the sequence(16 in this example)
    * **hidden size:** The vector dimension of each model input.(hidden size may very as 768 for common smaller model | 3072 for larger model.)
  

In [4]:
from transformers import AutoModel

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModel.from_pretrained(checkpoint)

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

In [5]:
outputs = model(**inputs)
print(outputs.last_hidden_state.shape)
# 2 batch size
# 16 sequence length
# 768 hidden size

torch.Size([2, 16, 768])


<img src="huggingface_model.png" width=800 height=400>

### 3. Model head: Making sense out of numbers
* The model heads take the high-dimensional vector of hidden states as input and project them onto a different dimension. 
* The output of the Transformer model is sent directly to the model head to be processed.
* There are many different architectures available in 🤗 Transformers, with each one designed around tackling a ***specific task***. Here is a non-exhaustive list:
    * ```*Model (retrieve the hidden states)```
    * ```*ForCausalLM```
    * ```*ForMaskedLM```
    * ```*ForMultipleChoice```
    * ```*ForQuestionAnswering```
    * ```*ForSequenceClassification```
    * ```*ForTokenClassification```

In [6]:
#model with a sequence classification head(classify sentence as positive | negative).
# so we won't use the AutoModel class, but AutoModelForSequenceClassification.

from transformers import AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
outputs = model(**inputs)

The history saving thread hit an unexpected error (OperationalError('database or disk is full')).History will not be written to the database.


In [8]:
print(outputs.logits.shape)
print(outputs.logits)

torch.Size([2, 2])
tensor([[-1.5607,  1.6123],
        [ 4.1692, -3.3464]], grad_fn=<AddmmBackward0>)


### 4. Postprocessing
<img src="./huggingface_post_processing.png" width=800 height=400>

In [9]:
import torch

predictions = torch.nn.functional.softmax(outputs.logits, dim=-1)
print(predictions)

tensor([[4.0195e-02, 9.5980e-01],
        [9.9946e-01, 5.4418e-04]], grad_fn=<SoftmaxBackward0>)


In [10]:
model.config.id2label

{0: 'NEGATIVE', 1: 'POSITIVE'}